# Challenge 3

In this challenge we will work on the `Orders.csv` data set in the previous [Subsetting and Descriptive Stats lab](../../lab-subsetting-and-descriptive-stats/your-code/main.ipynb). In your work you will apply the thinking process and workflow we showed you in Challenge 2.

You are serving as a Business Intelligence Analyst at the headquarter of an international fashion goods chain store. Your boss today asked you to do two things for her:

**First, identify two groups of customers from the data set.** The first group is **VIP Customers** whose **aggregated expenses** at your global chain stores are **above the 95th percentile** (aka. 0.95 quantile). The second group is **Preferred Customers** whose **aggregated expenses** are **between the 75th and 95th percentile**.

**Second, identify which country has the most of your VIP customers, and which country has the most of your VIP+Preferred Customers combined.**

# Q1: How to identify VIP & Preferred Customers?

We start by importing all the required libraries:

In [1]:
# import required libraries
import numpy as np
import pandas as pd

Next, import `Orders.csv` from the "subsetting" lab folder into a dataframe variable called `orders`. Print the head of `orders` to overview the data:

In [2]:
# enter your code here
orders = pd.read_csv('/Users/martin/desktop/Orders.csv')
orders.head()
#no sé porque no me deja leer el archivo de donde está en la carpeta de "datamex_082020/module-2/..." lo yengo que sacar al escritorioy ya me deja leerlo.

Unnamed: 0  InvoiceNo StockCode  year  month  day  hour  \
0           0     536365    85123A  2010     12    3     8   
1           1     536365     71053  2010     12    3     8   
2           2     536365    84406B  2010     12    3     8   
3           3     536365    84029G  2010     12    3     8   
4           4     536365    84029E  2010     12    3     8   

                           Description  Quantity          InvoiceDate  \
0   white hanging heart t-light holder         6  2010-12-01 08:26:00   
1                  white metal lantern         6  2010-12-01 08:26:00   
2       cream cupid hearts coat hanger         8  2010-12-01 08:26:00   
3  knitted union flag hot water bottle         6  2010-12-01 08:26:00   
4       red woolly hottie white heart.         6  2010-12-01 08:26:00   

   UnitPrice  CustomerID         Country  amount_spent  
0       2.55       17850  United Kingdom         15.30  
1       3.39       17850  United Kingdom         20.34  
2       2.75       17850  United Kingdom         22.00  
3       3.39       17850  United Kingdom         20.34  
4       3.39       17850  United Kingdom         20.34

In [5]:
Vip_customers = orders[orders['amount_spent'] > orders['amount_spent'].quantile(0.95)]
Vip_customers

Unnamed: 0  InvoiceNo StockCode  year  month  day  hour  \
26              26     536370     22728  2010     12    3     8   
27              27     536370     22727  2010     12    3     8   
34              34     536370     22326  2010     12    3     8   
40              40     536370     22900  2010     12    3     8   
46              46     536371     22086  2010     12    3     9   
...            ...        ...       ...   ...    ...  ...   ...   
397842      541827     581579     23313  2011     12    5    12   
397850      541835     581579     23581  2011     12    5    12   
397883      541868     581584     85038  2011     12    5    12   
397907      541892     581586     21217  2011     12    5    12   
397908      541893     581586     20685  2011     12    5    12   

                                Description  Quantity          InvoiceDate  \
26                alarm clock bakelike pink        24  2010-12-01 08:45:00   
27                alarm clock bakelike red         24  2010-12-01 08:45:00   
34      round snack boxes set of4 woodland         24  2010-12-01 08:45:00   
40          set 2 tea towels i love london         24  2010-12-01 08:45:00   
46          paper chain kit 50's christmas         80  2010-12-01 09:00:00   
...                                     ...       ...                  ...   
397842            vintage christmas bunting        15  2011-12-09 12:19:00   
397850               jumbo bag paisley park        40  2011-12-09 12:19:00   
397883      6 chocolate love heart t-lights        48  2011-12-09 12:25:00   
397907        red retrospot round cake tins        24  2011-12-09 12:49:00   
397908                doormat red retrospot        10  2011-12-09 12:49:00   

        UnitPrice  CustomerID         Country  amount_spent  
26           3.75       12583          France         90.00  
27           3.75       12583          France         90.00  
34           2.95       12583          France         70.80  
40           2.95       12583          France         70.80  
46           2.55       13748  United Kingdom        204.00  
...           ...         ...             ...           ...  
397842       4.95       17581  United Kingdom         74.25  
397850       1.79       17581  United Kingdom         71.60  
397883       1.85       13777  United Kingdom         88.80  
397907       8.95       13113  United Kingdom        214.80  
397908       7.08       13113  United Kingdom         70.80  

[19615 rows x 14 columns]

In [8]:
Preferred_customers = orders[(orders['amount_spent'] == ((orders['amount_spent'].quantile(0.95)) - (orders['amount_spent'].quantile(0.75))))]
Preferred_customers

Unnamed: 0  InvoiceNo StockCode  year  month  day  hour  \
2569          3750     536672     20749  2010     12    4    12   
4415          6586     536957     21523  2010     12    5    12   
5520          7899     537065     20750  2010     12    7    11   
7652         10043     537224     85048  2010     12    7    16   
9724         14050     537446     82486  2010     12    2     9   
...            ...        ...       ...   ...    ...  ...   ...   
387058      524590     580603     22802  2011     12    1    11   
392226      531885     580977     85048  2011     12    2    15   
393200      533713     581104     82484  2011     12    3    12   
394252      534787     581171     20750  2011     12    3    15   
395190      537466     581281     82484  2011     12    4    11   

                                Description  Quantity          InvoiceDate  \
2569             assorted colour mini cases         6  2010-12-02 12:10:00   
4415     doormat fancy font home sweet home         6  2010-12-03 12:43:00   
5520               red retrospot mini cases         6  2010-12-05 11:57:00   
7652    15cm christmas glass ball 20 lights         6  2010-12-05 16:24:00   
9724      wood s/3 cabinet ant white finish         6  2010-12-07 09:13:00   
...                                     ...       ...                  ...   
387058             faux fur chocolate throw         6  2011-12-05 11:31:00   
392226  15cm christmas glass ball 20 lights         6  2011-12-06 15:32:00   
393200    wood black board ant white finish         6  2011-12-07 12:11:00   
394252             red retrospot mini cases         6  2011-12-07 15:02:00   
395190    wood black board ant white finish         6  2011-12-08 11:36:00   

        UnitPrice  CustomerID         Country  amount_spent  
2569         7.95       14491  United Kingdom          47.7  
4415         7.95       17809  United Kingdom          47.7  
5520         7.95       12567          France          47.7  
7652         7.95       13174  United Kingdom          47.7  
9724         7.95       18055  United Kingdom          47.7  
...           ...         ...             ...           ...  
387058       7.95       16642  United Kingdom          47.7  
392226       7.95       16569  United Kingdom          47.7  
393200       7.95       13668  United Kingdom          47.7  
394252       7.95       12615          France          47.7  
395190       7.95       18109  United Kingdom          47.7  

[263 rows x 14 columns]

In [24]:
Preferred_customers['Country'].value_counts().head(1)

United Kingdom    208
Name: Country, dtype: int64

In [6]:
Vip_customers['Country'].value_counts().head(1)

United Kingdom    14739
Name: Country, dtype: int64

In [29]:
(Preferred_customers['Country'].value_counts().max() + Vip_customers['Country'].value_counts().max())

14947

United Kingdom es el país con más clientes 'Preferred', así como el país con más clientes 'Preferred' y 'Vip'
combinados con 14947

United Kingdom es el país con más clientes 'Preferred', así como el país con más clientes 'Preferred' y 'Vip'
combinados con 14947

We break down the main problem into several sub problems:

#### Sub Problem 1: How to aggregate the  `amount_spent` for unique customers?

#### Sub Problem 2: How to select customers whose aggregated `amount_spent` is in a given quantile range?

#### Sub Problem 3: How to label selected customers as "VIP" or "Preferred"?

*Note: If you want to break down the main problem in a different way, please feel free to revise the sub problems above.*

Now in the workspace below, tackle each of the sub problems using the iterative problem solving workflow. Insert cells as necessary to write your codes and explain your steps.

In [13]:
# your code here
orders.groupby('CustomerID')['amount_spent'].agg(['mean','std','sum','max','min','count']).head()

mean        std       sum      max       min  count
CustomerID                                                             
12346       77183.600000        NaN  77183.60  77183.6  77183.60      1
12347          23.681319  23.289902   4310.00    249.6      5.04    182
12348          57.975484  48.514857   1797.24    240.0     13.20     31
12349          24.076027  34.655913   1757.55    300.0      6.64     73
12350          19.670588   7.275538    334.40     40.0      8.50     17

In [33]:

sp2 = orders[orders['amount_spent']>orders['amount_spent'].quantile(0.5)]
sp2.head()

Unnamed: 0  InvoiceNo StockCode  year  month  day  hour  \
0           0     536365    85123A  2010     12    3     8   
1           1     536365     71053  2010     12    3     8   
2           2     536365    84406B  2010     12    3     8   
3           3     536365    84029G  2010     12    3     8   
4           4     536365    84029E  2010     12    3     8   

                           Description  Quantity          InvoiceDate  \
0   white hanging heart t-light holder         6  2010-12-01 08:26:00   
1                  white metal lantern         6  2010-12-01 08:26:00   
2       cream cupid hearts coat hanger         8  2010-12-01 08:26:00   
3  knitted union flag hot water bottle         6  2010-12-01 08:26:00   
4       red woolly hottie white heart.         6  2010-12-01 08:26:00   

   UnitPrice  CustomerID         Country  amount_spent  
0       2.55       17850  United Kingdom         15.30  
1       3.39       17850  United Kingdom         20.34  
2       2.75       17850  United Kingdom         22.00  
3       3.39       17850  United Kingdom         20.34  
4       3.39       17850  United Kingdom         20.34

In [32]:
sp2.groupby('CustomerID')['amount_spent'].agg(['mean','std','sum','max','min','count']).head()

mean        std       sum      max       min  count
CustomerID                                                             
12346       77183.600000        NaN  77183.60  77183.6  77183.60      1
12347          25.696000  24.152156   4111.36    249.6     12.48    160
12348          57.975484  48.514857   1797.24    240.0     13.20     31
12349          26.467460  36.766826   1667.45    300.0     12.48     63
12350          21.046667   6.568634    315.70     40.0     15.00     15

In [7]:
Vip_customers.groupby('CustomerID')['amount_spent'].agg(['mean','std','sum','max','min','count']).head()

mean         std       sum      max      min  count
CustomerID                                                             
12346       77183.600000         NaN  77183.60  77183.6  77183.6      1
12347         113.288571   61.683214    793.02    249.6     70.8      7
12348         124.250000   52.425485    994.00    240.0     69.6      8
12349         300.000000         NaN    300.00    300.0    300.0      1
12352         200.060000  129.058414   1000.30    376.5     80.0      5

In [9]:
Preferred_customers.groupby('CustomerID')['amount_spent'].agg(['mean','std','sum','max','min','count']).head()

mean  std   sum   max   min  count
CustomerID                                    
12359       47.7  0.0  95.4  47.7  47.7      2
12372       47.7  NaN  47.7  47.7  47.7      1
12398       47.7  NaN  47.7  47.7  47.7      1
12405       47.7  NaN  47.7  47.7  47.7      1
12413       47.7  NaN  47.7  47.7  47.7      1

Now we'll leave it to you to solve Q2 & Q3, which you can leverage from your solution for Q1:

# Q2: How to identify which country has the most VIP Customers?

# Q3: How to identify which country has the most VIP+Preferred Customers combined?

Provide your solution for Q2 below:

In [10]:
# your code here
Vip_customers['Country'].value_counts().head(1)

United Kingdom    14739
Name: Country, dtype: int64

q2: United Kingdom es el país con más 'Vip cuatomers'

In [11]:
(Preferred_customers['Country'].value_counts().head(1) + Vip_customers['Country'].value_counts().head(1))

United Kingdom    14947
Name: Country, dtype: int64

q3: United Kingdom es el pais con más 'Vip + Preferred customers' combinados.